snuba/query/processors/tags_promoter.py

```diff
@@ -0,0 +1,105 @@
+from typing import Mapping, NamedTuple
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.request.request_settings import RequestSettings
+
+
+class PromotedColumnsSpec(NamedTuple):
+    # The name of the key field in the mapping column
+    key_field: str
+    # The name of the value field in the mapping column
+    val_field: str
+    # The mapping between keys in the column and real columns
+    column_mapping: Mapping[str, str]
+
+
+MappingColumnPromotionSpec = Mapping[str, PromotedColumnsSpec]
+
+
+class MappingColumnPromoter(QueryProcessor):
+    """
+    Promotes expression to access the value of a mapping column by replacing them with
+    the corresponding promoted column.
+
+    Example: tags["myTag"] -> arrayElement("tags.value", indexOf("tags.key", "myTag"))
+     -> toString(promoted_MyTag)
+
+    This happens if there is a promoted_MyTag column in the storage that maps to the myTag tag.
+    """
+
+    def __init__(
+        self, columns: ColumnSet, mapping_spec: MappingColumnPromotionSpec
+    ) -> None:
+        # The ColumnSet of the dataset. Used to format promoted
+        # columns with the right type.
+        self.__columns = columns
+        # Maps nested column keys to real column names.
+        self.__spec = mapping_spec
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def transform_nested_column(exp: Expression) -> Expression:
+            # TODO: use the matcher system developed in #942 when merged to make this
+            # less verbose and put the code that generates this expression (in #936) in
+            # the same place as the matcher expression so they do not diverge.
+            #
+            # Now it identifies arrayElement("tags.value", indexOf("tags.key", "myTag"))
+            if not (
```

Comment:
With #942 this would look like:
```
pattern = FunctionCall(
    None,
    String("arrayElement"),
    (Column(None, Param("col_name", Any(str)), Param("value", Any(str)), None), None),
     FunctionCall(None, String("indexOf"), (
           Column(None, Param("col_name2", Any(str), Param("key", Any(str))), None),
           Literal(None, Param("tag_name", Any(str)))
     ))